src/filters.rs

```diff
@@ -0,0 +1,1115 @@
+use std::collections::{HashMap, HashSet};
+
+use zksync_basic_types::{H160, H256, U256, U64};
+use zksync_types::api::{BlockNumber, Log};
+use zksync_web3_decl::types::FilterChanges;
+
+/// Specifies a filter type
+#[derive(Debug, Clone, PartialEq)]
+pub enum FilterType {
+    /// A filter for block information
+    Block(BlockFilter),
+    /// A filter for log information.
+    /// This is [Box] to ensure the enum invariants are similar size.
+    Log(Box<LogFilter>),
+    /// A filter for pending transaction information
+    PendingTransaction(PendingTransactionFilter),
+}
+
+/// Specifies a filter that keeps track of new blocks
+#[derive(Debug, Default, Clone, PartialEq)]
+pub struct BlockFilter {
+    updates: Vec<H256>,
+}
+
+/// Specifies a filter that keeps track of new logs
+#[derive(Debug, Clone, PartialEq)]
+pub struct LogFilter {
+    from_block: BlockNumber,
+    to_block: BlockNumber,
+    addresses: Vec<H160>,
+    topics: [Option<HashSet<H256>>; 4],
+    updates: Vec<Log>,
+}
+
+impl LogFilter {
+    fn matches(&self, log: &Log, latest_block_number: U64) -> bool {
+        let from = match self.from_block {
```

Comment:
Should we create a helper function in `utils.rs` that returns `u64` given `BlockNumber`? I think we've used this in a few places now.

Reply:
I don't recall me adding this, can you perhaps give an example? In my case I needed to pass the `latest_block_number` since there's no way to map `latest` without this knowledge.

Reply:
Hmm, thought I needed to do this with gas estimation at some point, but looks like the only other places are for `get_transaction_count_by_number` and `get_block_by_number`.

Though it would be a slight refactor/optimization for `get_block_by_number`.

And since it would be a helper function, I think it would make sense to pass in `latest_block_number` instead of a `reader`.

Reply:
Either way, not necessary for this PR (resolving conversation)

Reply:
already committed it